### PR TITLE
Added support for innerRef

### PR DESCRIPTION
**Pull request by TriPSs**
Added support for innerRef, issue #212

### CODE REVIEW

File: src/createHoc.js [hunk 1]
```diff
@@ -210,14 +214,20 @@ export default (stylesOrCreator, InnerComponent, options = {}) => {
 
     render() {
       const {theme, dynamicSheet, classes} = this.state
+      const {innerRef, ...rest} = this.props
+
       const sheet = dynamicSheet || this.manager.get(theme)
       const props = {}
```
Comment (kof) [inline src/createHoc.js]: we can just rename ...rest to ...props and use it, so we don't need to use Object.assign later on as well

File: src/tests/inner-ref.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
+/* eslint-disable global-require, react/prop-types */
+import expect from 'expect.js'
+import React, {PureComponent} from 'react'
+import {spy} from 'sinon'
+
+describe('inner ref', () => {
+  it('should provide a ref on the inner component', () => {
+    const handleRef = spy()
+
+    /* eslint-disable react/no-multi-comp, react/prefer-stateless-function */
```
Comment (kof) [inline src/tests/inner-ref.js]: why not use a functional component?

File: src/createHoc.js [hunk 3]
```diff
@@ -210,14 +214,20 @@ export default (stylesOrCreator, InnerComponent, options = {}) => {
...
+
       if (injectMap.sheet) props.sheet = sheet
```
Comment (kof) [inline src/createHoc.js]: in that case you will have to add an additional condition here in order to not override

File: src/tests/inner-ref.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
+/* eslint-disable global-require, react/prop-types */
+import expect from 'expect.js'
+import React, {PureComponent} from 'react'
+import {spy} from 'sinon'
+
+describe('inner ref', () => {
```
Comment (kof) [inline src/tests/inner-ref.js]: this test belongs to injectSheet.test.js

Comment (TriPSs) [reply src/tests/inner-ref.js]: Yes, i did not see that test file!

File: src/tests/inner-ref.js (hunk 2 above)
Comment (TriPSs) [reply src/tests/inner-ref.js]: Because of:
image

File: src/createHoc.js (hunk 3 above)
Comment (TriPSs) [reply src/createHoc.js]: Override what exactly?

Comment (kof) [reply src/createHoc.js]: In other words we need to make sure that props passed by the user override hoc props. At least that was the logic so far.

Comment (TriPSs) [reply src/createHoc.js]: Makes sense.

Think we have to do the same for theme?
```js
      if (injectMap.sheet && !props.sheet) props.sheet = sheet
      if (isThemingEnabled && injectMap.theme && !props.theme) props.theme = theme
```

Comment (kof) [reply src/createHoc.js]: yep

**Comment by HenriBeck** [comment]
@TriPSs can you add documentation for it?

**Comment by kof** [comment]
I am already on it

**Comment by kof** [comment]
Ok, released 8.4.0, also I realised that you mised up inner component and inner element in the tests. What react calls refs are references to component instances aka elements. References to components are normal javascript references.

**Comment by kof** [comment]
in any case good work on landing this one @TriPSs and thanks @HenriBeck for looking over. Please test the release if something is wrong I am back in 9h.

**Comment by HenriBeck** [comment]
https://reactjs.org/docs/forwarding-refs.html
Now that react 16.3 is released there is a supported way of forwarding refs.

**Comment by kof** [comment]
Do we need any changes to support forwardRef?

**Comment by HenriBeck** [comment]
Yes, we either need to release a version only for `react@^16.3` or we would need to check if `forwardRef` exists on the `React` object and only use it when it's actually supported.

If we were to release a separate version for `react@^16.3` we could also use the new lifecycle methods and deprecate the old ones maybe.

**Comment by kof** [comment]
If we assume that `React. forwardRef`  will be used by the end consumer, it seems that we don't need to do anything, it will just work (from looking at the docs).

**Pull request merged**